### PR TITLE
BUG: Do not allow boolean values for indexing with .loc without a boolean index.

### LINKED ISSUE #20432: BUG: DataFrame/Series.loc improperly allows lookups of boolean labels/slices
#### Code Sample, a copy-pastable example if possible
Basic example of the issue, specific to `TimedeltaIndex`, xref https://github.com/pandas-dev/pandas/pull/20408#discussion_r175627548
```python
In [2]: s = pd.Series(list('abcde'), pd.timedelta_range(0, 4, freq='ns'))

In [3]: s.loc[True]
Out[3]: 'b'

In [4]: s.loc[False:True]
Out[4]:
00:00:00           a
00:00:00.000000    b
Freq: N, dtype: object
```

Indexing with both boolean labels and slices was successful, which doesn't seem right.

I investigated this same behavior across various index types for both `Series` and `DataFrame`, and produced the summary below.

**Summary**
- 'raises' column indicates if the indexing operation raised an exception
-  'exception' column indicates the type of exception raised
<details>

```
                                  raises  exception
CategoricalIndex DataFrame label    True   KeyError
                           slice    True   KeyError
                 Series    label   False        NaN
                           slice    True   KeyError
DatetimeIndex    DataFrame label   False        NaN
                           slice   False        NaN
                 Series    label   False        NaN
                           slice   False        NaN
Float64Index     DataFrame label   False        NaN
                           slice   False        NaN
                 Series    label   False        NaN
                           slice   False        NaN
Index            DataFrame label   False        NaN
                           slice   False        NaN
                 Series    label   False        NaN
                           slice   False        NaN
Int64Index       DataFrame label    True   KeyError
                           slice   False        NaN
                 Series    label    True   KeyError
                           slice   False        NaN
IntervalIndex    DataFrame label    True  TypeError
                           slice    True  TypeError
                 Series    label    True  TypeError
                           slice    True  TypeError
MultiIndex       DataFrame label    True   KeyError
                           slice    True   KeyError
                 Series    label    True   KeyError
                           slice    True   KeyError
PeriodIndex      DataFrame label    True   KeyError
                           slice   False        NaN
                 Series    label    True   KeyError
                           slice   False        NaN
RangeIndex       DataFrame label    True   KeyError
                           slice   False        NaN
                 Series    label    True   KeyError
                           slice   False        NaN
TimedeltaIndex   DataFrame label   False        NaN
                           slice   False        NaN
                 Series    label   False        NaN
                           slice   False        NaN
UInt64Index      DataFrame label    True   KeyError
                           slice   False        NaN
                 Series    label    True   KeyError
                           slice   False        NaN
```
</details>

<br />

**Code to produce summary**
<details>

```python
indexes = [
    pd.RangeIndex(4),
    pd.Int64Index(range(4)),
    pd.UInt64Index(range(4)),
    pd.Float64Index(range(4)),
    pd.CategoricalIndex(range(4)),
    pd.date_range(0, periods=4, freq='ns'),
    pd.timedelta_range(0, periods=4, freq='ns'),
    pd.interval_range(0, periods=4),
    pd.Index([0, 1, 2, 3], dtype=object),
    pd.MultiIndex.from_product([[0, 1], [0, 1]]),
    pd.period_range('2018Q1', freq='Q', periods=4),  # need better example here
]

result = {}
for index in indexes:
    index_name = type(index).__name__
    s = pd.Series(list('abcd'), index=index)
    for obj in (s, s.to_frame()):
        obj_name = type(obj).__name__

        # check single label
        key = (index_name, obj_name, 'label')
        try:
            obj.loc[True]
            result[key] = {'raises': False}
        except Exception as e:
            result[key] = {'raises': True, 'exception': type(e).__name__}

        # check slice
        key = (index_name, obj_name, 'slice')
        try:
            obj.loc[False:True]
            result[key] = {'raises': False}
        except Exception as e:
            result[key] = {'raises': True, 'exception': type(e).__name__}

result = pd.DataFrame.from_dict(result, orient='index')
```
</details>

#### Expected Output
I'd generally expect all of these operations to raise a `KeyError`, which a couple potential exceptions:

- I'd be open to an argument for numeric indexes casting to integer equivalent.  Seems like this should at least be consistent for labels vs slices, which it is not right now.
- Maybe we should allow conversion for the `object` dtype `Index`?

### PR DESCRIPTION
**Pull request by DriesSchaumont**
- [x] closes #20432
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -961,12 +961,17 @@ class _LocIndexer(_LocationIndexer):
...
     def _validate_key(self, key, axis: int):
-
         # valid for a collection of labels (we check their presence later)
         # slice of labels (where start-end in labels)
         # slice of integers (only if in the labels)
-        # boolean
-        pass
+        # boolean not in slice and with boolean index
+        if not is_bool_dtype(self.obj.index) and isinstance(key, bool):
+            raise TypeError("Boolean label can not be used without a boolean index")
```
Comment (phofl) [inline pandas/core/indexing.py:969]: Why not KeyError?

could you run asv? Iirc is_bool_dtype is not completly cheap, but might be wrong

File: pandas/tests/indexing/test_loc.py [hunk 2]
```diff
@@ -2085,6 +2088,60 @@ def test_loc_getitem_slice_columns_mixed_dtype(self):
...
+        ],
+    )
+    def test_loc_bool_incompatible_index_raises(self, index, frame_or_series):
+        # GH20432
+        message = "Boolean label can not be used without a boolean index"
+        obj = frame_or_series(range(4), index=index)
+        with pytest.raises(TypeError, match=message):
+            obj.loc[True]
+
+    @pytest.mark.parametrize("index", [Index([True, False], dtype="boolean")])
```
Comment (phofl) [inline pandas/tests/indexing/test_loc.py:2115]: Remove this parametrization

File: pandas/tests/indexing/test_loc.py [hunk 3]
```diff
@@ -2085,6 +2088,60 @@ def test_loc_getitem_slice_columns_mixed_dtype(self):
...
+            pd.Float64Index(range(4)),
+            CategoricalIndex(range(4)),
+            date_range(0, periods=4, freq="ns"),
+            timedelta_range(0, periods=4, freq="ns"),
+            pd.interval_range(0, periods=4),
+            Index([0, 1, 2, 3], dtype=object),
+            Index([True, True, False, False], dtype=object),
+            MultiIndex.from_product([[0, 1], [0, 1]]),
+            pd.period_range("2018Q1", freq="Q", periods=4),
+        ],
```
Comment (phofl) [inline pandas/tests/indexing/test_loc.py:2135]: We probably have a fixture for the different index types?

Comment (jreback) [reply pandas/tests/indexing/test_loc.py:2135]: yep fixture is 'index'!

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 1 above)
Comment (DriesSchaumont) [reply pandas/core/indexing.py:969]:
```
       before           after         ratio
     [1367cacd]       [ee07ee02]
                      <fix-20432>
+         323±7μs         553±70μs     1.71  indexing.AssignTimeseriesIndex.time_frame_assign_timeseries_index
+        52.1±1μs       71.0±0.4μs     1.36  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+      38.4±0.9μs       49.4±0.8μs     1.29  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
+      19.6±0.6μs       24.0±0.2μs     1.22  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     6.82±0.09ms       8.18±0.9ms     1.20  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
+      49.8±0.7μs       59.7±0.3μs     1.20  indexing.DataFrameNumericIndexing.time_loc
+      20.3±0.5μs       23.8±0.6μs     1.18  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
+        66.0±1μs         76.8±2μs     1.16  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monotonic_inc')
+        91.9±2μs        106±0.8μs     1.16  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monotonic_inc')
+        55.2±2μs         61.7±1μs     1.12  indexing.NumericSeriesIndexing.time_getitem_scalar(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monotonic_inc')
+         104±2μs          115±2μs     1.11  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
-      2.27±0.3ms      1.97±0.02ms     0.87  indexing.IntervalIndexing.time_getitem_scalar
-        588±10μs          493±7μs     0.84  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
-       334±100μs          266±5μs     0.80  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Float64Engine'>, <class 'numpy.float64'>), 'monotonic_decr')
-       320±100μs          228±3μs     0.71  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int32Engine'>, <class 'numpy.int32'>), 'monotonic_decr')
-        211±40μs          151±1μs     0.71  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-     1.98±0.06ms      1.33±0.05ms     0.67  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
```
That is a pretty big penalty indeed. What I could do is swap the statements so `is_bool_dtype` will only be called for boolean keys. In that case the performance reduction should only occur for boolean keys. Or is there a faster way to do this check? As far as I know there is no boolean index type? 

Regarding why I chose `TypeError` over `KeyError`: I think it depends if you think the issue is the key or the index? I have no objection to changing it. But at least for boolean slices, TypeError seems to be the correct error type.

File: pandas/tests/indexing/test_loc.py (hunk 3 above)
Comment (DriesSchaumont) [reply pandas/tests/indexing/test_loc.py]: Solved!

File: pandas/tests/indexing/test_loc.py (hunk 2 above)
Comment (DriesSchaumont) [reply pandas/tests/indexing/test_loc.py]: Done!

Comment (DriesSchaumont) [reply pandas/tests/indexing/test_loc.py]: Should I integrate this test in the one above it?

File: pandas/core/indexing.py (hunk 1 above)
Comment (phofl) [reply pandas/core/indexing.py:969]: Hm we are using mostly KeyError, so I would prefer this here. 

Yeah checking the key first should help. I do not know a faster way

File: pandas/tests/indexing/test_loc.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/indexing/test_loc.py]: Is ok like this imo

**Comment by phofl** [comment]
Please add whatsnew

File: pandas/core/indexing.py (hunk 1 above)
Comment (DriesSchaumont) [reply pandas/core/indexing.py:969]: Changed to KeyError and switched the boolean statements.

**Comment by DriesSchaumont** [comment]
> Please add whatsnew

Added.

File: pandas/core/indexing.py (hunk 1 above)
Comment (phofl) [reply pandas/core/indexing.py:969]: Could you run asvs again?

File: pandas/core/dtypes/common.py [hunk 4]
```diff
@@ -1386,7 +1386,7 @@ def is_bool_dtype(arr_or_dtype) -> bool:
         # we don't have a boolean Index class
         # so its object, we need to infer to
         # guess this
-        return arr_or_dtype.is_object and arr_or_dtype.inferred_type == "boolean"
+        return arr_or_dtype.is_object() and arr_or_dtype.inferred_type == "boolean"
```
Comment (jreback) [inline pandas/core/dtypes/common.py:1389]: wow! did this change anything? (w/o the rest of the patch)

File: pandas/core/indexing.py (hunk 1 above)
Comment (DriesSchaumont) [reply pandas/core/indexing.py]:
```
       before           after         ratio
     [cc84a238]       [b90967d7]
     <fix-40014~1>       <fix-20432>
+         113±3μs         203±60μs     1.79  indexing.DataFrameStringIndexing.time_boolean_rows
+      3.80±0.1μs       4.71±0.4μs     1.24  indexing.DataFrameStringIndexing.time_getitem_scalar
+     2.99±0.06ms      3.34±0.04ms     1.12  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      42.9±0.3μs         47.3±1μs     1.10  indexing.NumericSeriesIndexing.time_iloc_list_like(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
-      46.0±0.8μs       41.3±0.3μs     0.90  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
-        277±10μs          248±5μs     0.89  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Float32Engine'>, <class 'numpy.float32'>), 'non_monotonic')
-        453±90μs          392±1μs     0.86  indexing.Take.time_take('int')
-        3.93±2μs      2.94±0.05μs     0.75  indexing_engines.ObjectEngineIndexing.time_get_loc('monotonic_incr')
```

File: pandas/core/dtypes/common.py (hunk 4 above)
Comment (DriesSchaumont) [reply pandas/core/dtypes/common.py:1389]: Afaik, no other tests were failing after I changed this. I am running all tests on master with only this change just to make sure.

Comment (DriesSchaumont) [reply pandas/core/dtypes/common.py:1389]: Perhaps I should do a coverage check to see which tests covers this line.

Comment (DriesSchaumont) [reply pandas/core/dtypes/common.py:1389]: Tests were green.

**Comment by phofl** [comment]
CI failure seems related

**Comment by DriesSchaumont** [comment]
> CI failure seems related

Yes, indeed. I could fix most of the failing tests with specifying the dtypes. But for some reason specifying dtype="boolean" does not remove the warning. This is a separate issue though.

**Comment by DriesSchaumont** [comment]
> > CI failure seems related
> 
> Yes, indeed. I could fix most of the failing tests with specifying the dtypes. But for some reason specifying dtype="boolean" does not remove the warning. This is a separate issue though.

I must have been tired. dtype="boolean" was not passed to the df or series, but to the index. Fixed and expect CI to be green.

File: doc/source/whatsnew/v1.3.0.rst [hunk 5]
```diff
@@ -642,6 +642,7 @@ Indexing
 - Bug in setting ``numpy.timedelta64`` values into an object-dtype :class:`Series` using a boolean indexer (:issue:`39488`)
 - Bug in setting numeric values into a into a boolean-dtypes :class:`Series` using ``at`` or ``iat`` failing to cast to object-dtype (:issue:`39582`)
 - Bug in :meth:`DataFrame.loc.__setitem__` when setting-with-expansion incorrectly raising when the index in the expanding axis contains duplicates (:issue:`40096`)
+- Bug in :meth:`DataFrame.loc` incorrectly allowing the lookup of boolean labels and slices (:issue:`20432`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.3.0.rst:645]: something like incorrectly matching non-boolean index elements?

File: pandas/core/indexing.py [hunk 6]
```diff
@@ -961,12 +961,17 @@ class _LocIndexer(_LocationIndexer):
...
     def _validate_key(self, key, axis: int):
-
         # valid for a collection of labels (we check their presence later)
         # slice of labels (where start-end in labels)
         # slice of integers (only if in the labels)
-        # boolean
-        pass
+        # boolean not in slice and with boolean index
+        if isinstance(key, bool) and not is_bool_dtype(self.obj.index):
+            raise KeyError("Boolean label can not be used without a boolean index")
```
Comment (phofl) [inline pandas/core/indexing.py:969]: Could we add the key here?

File: pandas/core/indexing.py [hunk 7]
```diff
@@ -961,12 +961,17 @@ class _LocIndexer(_LocationIndexer):
...
+
+        if isinstance(key, slice) and (
+            isinstance(key.start, bool) or isinstance(key.stop, bool)
+        ):
+            raise TypeError("Boolean values can not be used in a slice")
```
Comment (phofl) [inline pandas/core/indexing.py:974]: Same here maybe?

File: pandas/tests/indexing/test_loc.py [hunk 8]
```diff
@@ -2116,6 +2116,29 @@ def test_loc_getitem_slice_columns_mixed_dtype(self):
         tm.assert_frame_equal(df.loc[:, 1:], expected)
 
 
+class TestLocBooleanLabelsAndSlices(Base):
+    def test_loc_bool_incompatible_index_raises(self, index, frame_or_series):
```
Comment (phofl) [inline pandas/tests/indexing/test_loc.py:2120]: To be on the safe side: Could you parametrize over True and False?

File: pandas/tests/indexing/test_loc.py [hunk 9]
```diff
@@ -2116,6 +2116,29 @@ def test_loc_getitem_slice_columns_mixed_dtype(self):
...
+        # GH20432
+        message = "Boolean label can not be used without a boolean index"
+        if index.inferred_type != "boolean":
+            obj = frame_or_series(index=index, dtype="object")
+            with pytest.raises(KeyError, match=message):
+                obj.loc[True]
+
+    def test_loc_bool_should_not_raise(self, frame_or_series):
```
Comment (phofl) [inline pandas/tests/indexing/test_loc.py:2128]: Same here

**Comment by phofl** [comment]
Small comment, otherwise lgtm, pls merge master too

File: doc/source/whatsnew/v1.3.0.rst (hunk 5 above)
Comment (DriesSchaumont) [reply doc/source/whatsnew/v1.3.0.rst]: Done

File: pandas/core/indexing.py (hunk 6 above)
Comment (DriesSchaumont) [reply pandas/core/indexing.py]: Done!

File: pandas/core/indexing.py (hunk 7 above)
Comment (DriesSchaumont) [reply pandas/core/indexing.py]: Added!

File: pandas/tests/indexing/test_loc.py (hunk 8 above)
Comment (DriesSchaumont) [reply pandas/tests/indexing/test_loc.py]: Added!

File: pandas/tests/indexing/test_loc.py (hunk 9 above)
Comment (DriesSchaumont) [reply pandas/tests/indexing/test_loc.py]: Fixed!

**Comment by phofl** [comment]
lgtm cc @jreback @jbrockmendel

**Review by phofl — APPROVED** [review]

**Comment by jreback** [comment]
@DriesSchaumont did you post the updated asv's?

**Comment by phofl** [comment]
https://github.com/pandas-dev/pandas/pull/40726#discussion_r606390899

Here

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@phofl pls merge!

**Comment by phofl** [comment]
thanks @DriesSchaumont

**Pull request merged**